# Lab 2.10.4: Analysis—All Greens Franchise 
For this problem, we are going to use the All Greens Franchise data set. Download All Greens Franchise data set. It contains 27 observations about All Greens sales that have five predictor variables apart from the annual net sales figure. Use this dataset and perform the following sets of analyses:<p>
1. Find the correlation of annual net sales with money spent on advertising and the number of competitors in the area. Provide your one-sentence interpretation.
2. Build a regression model to predict the annual net sales figure using the other five columns in the data set. Report your model (line equation).
2. Now use gradient descent to learn the regression model. Report your model (gradients) along with the learning rate(s) you used and the cost(s).

In [11]:
import pandas as pd
df = pd.read_excel('LAB 2.10.4_ All greens franchise.xls')
print(df.columns)
display(df)

Index(['X1', 'X2', 'X3', 'X4', 'X5', 'X6'], dtype='object')


X1   X2   X3    X4         X5  X6
0   231.0  3.0  294   8.2   8.200000  11
1   156.0  2.2  232   6.9   4.100000  12
2    10.0  0.5  149   3.0   4.300000  15
3   519.0  5.5  600  12.0  16.100000   1
4   437.0  4.4  567  10.6  14.100000   5
5   487.0  4.8  571  11.8  12.700000   4
6   299.0  3.1  512   8.1  10.100000  10
7   195.0  2.5  347   7.7   8.400000  12
8    20.0  1.2  212   3.3   2.100000  15
9    68.0  0.6  102   4.9   4.700000   8
10  570.0  5.4  788  17.4  12.300000   1
11  428.0  4.2  577  10.5  14.000000   7
12  464.0  4.7  535  11.3  15.000000   3
13   15.0  0.6  163   2.5   2.500000  14
14   65.0  1.2  168   4.7   3.300000  11
15   98.0  1.6  151   4.6   2.700000  10
16  398.0  4.3  342   5.5  16.000000   4
17  161.0  2.6  196   7.2   6.300000  13
18  397.0  3.8  453  10.4  13.900000   7
19  497.0  5.3  518  11.5  16.299999   1
20  528.0  5.6  615  12.3  16.000000   0
21   99.0  0.8  278   2.8   6.500000  14
22    0.5  1.1  142   3.1   1.600000  12
23  347.0  3.6  461   9.6  11.300000   6
24  341.0  3.5  382   9.8  11.500000   5
25  507.0  5.1  590  12.0  15.700000   0
26  400.0  8.6  517   7.0  12.000000   8

In [12]:
# There is no information on which column is which, so guessing
df = df.rename(columns={'X1': 'sales', 'X3': 'ad_dollars', 'X6': 'num_competitors', 'X2': 'x3', 'X4': 'x4', 'X5': 'x5'})
display(df)

sales   x3  ad_dollars    x4         x5  num_competitors
0   231.0  3.0         294   8.2   8.200000               11
1   156.0  2.2         232   6.9   4.100000               12
2    10.0  0.5         149   3.0   4.300000               15
3   519.0  5.5         600  12.0  16.100000                1
4   437.0  4.4         567  10.6  14.100000                5
5   487.0  4.8         571  11.8  12.700000                4
6   299.0  3.1         512   8.1  10.100000               10
7   195.0  2.5         347   7.7   8.400000               12
8    20.0  1.2         212   3.3   2.100000               15
9    68.0  0.6         102   4.9   4.700000                8
10  570.0  5.4         788  17.4  12.300000                1
11  428.0  4.2         577  10.5  14.000000                7
12  464.0  4.7         535  11.3  15.000000                3
13   15.0  0.6         163   2.5   2.500000               14
14   65.0  1.2         168   4.7   3.300000               11
15   98.0  1.6         151   4.6   2.700000               10
16  398.0  4.3         342   5.5  16.000000                4
17  161.0  2.6         196   7.2   6.300000               13
18  397.0  3.8         453  10.4  13.900000                7
19  497.0  5.3         518  11.5  16.299999                1
20  528.0  5.6         615  12.3  16.000000                0
21   99.0  0.8         278   2.8   6.500000               14
22    0.5  1.1         142   3.1   1.600000               12
23  347.0  3.6         461   9.6  11.300000                6
24  341.0  3.5         382   9.8  11.500000                5
25  507.0  5.1         590  12.0  15.700000                0
26  400.0  8.6         517   7.0  12.000000                8

## #1

In [13]:
import statsmodels.formula.api as smf

model = smf.ols('sales ~ ad_dollars + num_competitors', data=df)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.958
Model:                            OLS   Adj. R-squared:                  0.954
Method:                 Least Squares   F-statistic:                     271.2
Date:                Sat, 12 Apr 2025   Prob (F-statistic):           3.35e-17
Time:                        22:03:14   Log-Likelihood:                -137.09
No. Observations:                  27   AIC:                             280.2
Df Residuals:                      24   BIC:                             284.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept         182.7286     47.615     

## ad_dollars
For every one-unit increase in advertising dollars spent, holding the number of competitors constant, sales are estimated to increase by approximately 0.60 units. This indicates a positive relationship between advertising expenditure and sales.

## num_competitors
For every additional competitor, holding advertising dollars constant, sales are estimated to decrease by approximately 16.77 units. This indicates a negative relationship between the number of competitors and sales.

## #2

$\text{Annual Net Sales} = \beta_0 + \beta_1\text{ad_dollars} + \beta_2\text{num_competitors} + \beta_3\text{x_3} + \beta_4\text{x_4} + \beta_5\text{x_5}$

In [14]:
model = smf.ols('sales ~ ad_dollars + num_competitors + x3 + x4 + x5', data=df)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.993
Model:                            OLS   Adj. R-squared:                  0.992
Method:                 Least Squares   F-statistic:                     611.6
Date:                Sat, 12 Apr 2025   Prob (F-statistic):           5.40e-22
Time:                        22:03:28   Log-Likelihood:                -112.43
No. Observations:                  27   AIC:                             236.9
Df Residuals:                      21   BIC:                             244.6
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept         -18.8594     30.150     

$\text{Annual Net Sales} = -18.8594 + 0.1746*\text{ad_dollars} - 5.3110*\text{num_competitors} + 16.2016*\text{x_3} + 11.5263*\text{x_4} + 13.5803*\text{x_5}$

In [17]:
def gradient_descent(X,y,theta,learning_rate=0.01,iterations=100):
    '''
    X    = Matrix of X with added bias units
    y    = Vector of Y
    theta=Vector of thetas np.random.randn(j,1)
    learning_rate 
    iterations = no of iterations
    Returns the final theta vector and array of cost history over no of iterations
    '''
    n = len(y)
    num_theta = X.shape[1]
    cost_history = np.zeros(iterations)
   # theta_history = np.zeros((iterations,2))
    theta_history = np.zeros((iterations,num_theta))
    for it in range(iterations):
        prediction = np.dot(X,theta)
        theta = theta -(1/n)*learning_rate*( X.T.dot((prediction - y)))
        theta_history[it,:] =theta.T
        cost_history[it]  = cost_fun(theta,X,y)
    return theta, cost_history, theta_history    

def cost_fun(theta,X,y):
    '''Calculates the cost for given X and Y. The following shows and example of a single dimensional X
    theta = Vector of thetas 
    X     = Row of X's np.zeros((2,j))
    y     = Actual y's np.zeros((2,1)) only for simple regression
    where:
        j is the no of features
    '''
    n = len(y)
    predictions = X.dot(theta)
    cost = (1/2*n) * np.sum((predictions-y)**2)
    return cost

In [31]:
import numpy as np
# Prepare the data
X = df[['ad_dollars', 'num_competitors', 'x3', 'x4', 'x5']].values
y = df['sales'].values
m = len(y)
n = X.shape[1]  # Number of features (Speed and Angle)

# Initialize parameters (coefficients)
learning_rate = 0.00001
iterations = 100
theta = np.zeros(n + 1)  # [intercept, slope_speed, slope_angle]

# Add bias term (for the intercept) to X
X_b = np.c_[np.ones((m, 1)), X]
# Train the model using Gradient Descent
theta_final, cost_history, theta_history = gradient_descent(X_b, y, theta, learning_rate, iterations)

# Get the learned coefficients
intercept_gd = theta_final[0]
slope_ad_gd = theta_final[1]
slope_comp_gd = theta_final[2]
slope_x3_gd = theta_final[3]
slope_x4_gd = theta_final[4]
slope_x5_gd = theta_final[5]

print("Gradient Descent Model Coefficients:")
print(f"Intercept: {intercept_gd:.4f}")
print(f"Slope (Ad Dollars): {slope_ad_gd:.4f}")
print(f"Slope (Num Competitors): {slope_comp_gd:.4f}")

Gradient Descent Model Coefficients:
Intercept: -0.0133
Slope (Ad Dollars): 0.7794
Slope (Num Competitors): -0.3592
